1-Draft/RFCNNNN-Supports-Parallel.md

```diff
@@ -0,0 +1,316 @@
+        # the array enclosures can be omitted. If the command should support
+        # parallel execution in all parameter sets, the command author can
+        # simply drop the equals operator and the values after it.
+        SupportsParallel=@('FilePathRunspace','Session','VMId')
+    )]
+    param(
+        # Define parameters here
+
+        # NEW! Any parameter sets that support parallel invocation would
+        # automatically receive additional common parameters. Like other common
+        # parameters, these could be used in Intellisense and tab completion,
+        # but they wouldn't be defined in a param block. They would be assigned
+        # automatically to the appropriate parameter sets by PowerShell.
+        # [-Parallel] : A switch used to run a command in parallel with the
+        #               default configuration for parallel command execution.
+        #               This parameter is not required if any of the following
+        #               parameters are used in the invocation.
+        # [-ThreadLimit <int>] : An int used to define the maximum number of
+        #                        threads to use during the parallel command
+        #                        execution (default value depends on system
+        #                        configuration)
+        # [-TimeoutSecs <int>] : An int used to define the maximum time allowed
+        #                        for all threads to complete. This parameter
+        #                        cannot be used in conjunction with the next
+        #                        two parameters.
+
+        # NOTE! [-AsJob] and [-JobName <string>] are intentionally excluded
+        # from this list because that work must be done external to the
+        # pipeline. The reasoning for this decision is explained in the
+        # "Alternate Proposals and Considerations" section below, under the
+        # subheading of "Running a parallelized command asynchronously".
+    )
+    begin {
+        # No changes here, this would be invoked synchronously in the current
+        # session.
+    }
+    process {
+        # NEW! The way this is invoked depends on whether or not the parameter
+        # set that was used supports parallelization or not.
+        #
+        # For a parameter set that does not support parallelization, this would
+        # be invoked synchronously.
+        #
+        # For a parameter set that supports parallelization, this would be
+        # invoked in a thread job, and the command would wait for all thread
+        # jobs to complete before continuing with end (and dispose) blocks.
+
+        # NEW! Unlike Start-Job commands where you must use the "using:" prefix
+        # on variables to access variables that are defined outside of the
+        # scope of the script block, all variables that are defined in the
+        # current scope would be accessible when the command is invoked with
```

Comment:
Can you clarify what you mean by "the current scope" here?  Do you mean the SessionState at the Invoke-Command call site?  If so, how do envision variables defined at the call site will be resolved from the perspective of `process{}`?  Would the SessionState to which the `process{}` scriptblock is bound be a "child" of the SessionState from which Invoke-Command was invoked?  (This would be similar to the way module SessionStates often have the "global" SessionState as a parent.)

_(For readers unfamiliar with PowerShell's SessionState scoped variable search "rules", [@lzybkr provides an explanation in #6139](https://github.com/PowerShell/PowerShell/issues/6139#issuecomment-364262570).  The important thing to understand is that those "rules" result in variable resolution behavior that is substantially different from C#.)_